### PR TITLE
Improved procedural surfaces, textures and additional standard uniforms

### PR DESCRIPTION
**Pull request by jherico**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: cf2850ccdcdca0403d14306a31a48df1a4cc0793 available at: https://gist.github.com/57a5b99be37ee94e1423<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.dmg

File: libraries/procedural/src/procedural/Procedural.cpp [hunk 1]
```diff
@@ -164,69 +194,169 @@ void Procedural::prepare(gpu::Batch& batch, const glm::vec3& size) {
...
+        }
+        QJsonValue value = _parsedUniforms[key];
+        if (value.isDouble()) {
+            float v = value.toDouble();
+            _uniforms.push_back([=](gpu::Batch& batch) {
+                batch._glUniform1f(slot, v);
+            });
+        } else if (value.isArray()) {
+            auto valueArray = value.toArray();
+            switch (valueArray.size()) {
```
Comment (ZappoMan) [inline libraries/procedural/src/procedural/Procedural.cpp]: coding standard: this switch/case statement needs to be formatted correctly.

**Comment by hifi-gustavo** [comment]
Build console output for commit: b30c78bf69e15f9df2a23f15538ca34f408b75ed available at: https://gist.github.com/57a5b99be37ee94e1423<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bfc4402af7458874d61608b5e528abd89bc37d32 available at: https://gist.github.com/57a5b99be37ee94e1423<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.dmg

**Comment by jherico** [comment]
RAGH!!!  You're making me start up my VM.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d41987141ddb0d9e2d2d864b8dae73643566956f available at: https://gist.github.com/57a5b99be37ee94e1423<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.dmg

File: libraries/procedural/src/procedural/Procedural.cpp (hunk 1 above)
Comment (jherico) [inline libraries/procedural/src/procedural/Procedural.cpp]: fixed

**Comment by hifi-gustavo** [comment]
Build console output for commit: e382da6ff65b48becd0abfea03e47e6d0187ca8d available at: https://gist.github.com/57a5b99be37ee94e1423<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 83c9ebf06cfe00e3a0226484dfd594263a40e318 available at: https://gist.github.com/57a5b99be37ee94e1423<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c8aea675057a0ed74193a1673ca054eda58ea394 available at: https://gist.github.com/57a5b99be37ee94e1423<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5946.dmg

File: libraries/gpu/src/gpu/Batch.h [hunk 2]
```diff
@@ -156,24 +156,24 @@ class Batch {
...
-    public:
-        uint  _count{ 0 };
-        uint  _instanceCount{ 0 };
-        uint  _firstIndex{ 0 };
-        uint  _baseInstance{ 0 };
+    
+    // multi command desctription for multiDrawIndexedIndirect
+    class DrawIndirectCommand {
+    public:
+        uint  _count{ 0 };
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/Batch.h]: coding standard: public members should not have "_" prefix

File: libraries/gpu/src/gpu/Batch.h [hunk 3]
```diff
@@ -156,24 +156,24 @@ class Batch {
...
-        uint  _count{ 0 };
-        uint  _instanceCount{ 0 };
-        uint  _firstIndex{ 0 };
-        uint  _baseVertex{ 0 };
-        uint  _baseInstance{ 0 };
+
+    // multi command desctription for multiDrawIndexedIndirect
+    class DrawIndexedIndirectCommand {
+    public:
+        uint  _count{ 0 };
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/Batch.h]: coding standard: public members should not have "_" prefix

File: libraries/procedural/src/procedural/Procedural.cpp [hunk 4]
```diff
@@ -164,69 +194,183 @@ void Procedural::prepare(gpu::Batch& batch, const glm::vec3& size) {
...
+            batch._glUniform(_standardUniformSlots[TIME], time);
+        });
+    }
+
+    if (gpu::Shader::INVALID_LOCATION != _standardUniformSlots[DATE]) {
+        _uniforms.push_back([=](gpu::Batch& batch) {
+            QDateTime now = QDateTime::currentDateTimeUtc();
+            QDate date = now.date();
+            QTime time = now.time();
+            vec4 v;
```
Comment (ZappoMan) [inline libraries/procedural/src/procedural/Procedural.cpp]: coding standard: variables should be clearly named

File: libraries/procedural/src/procedural/ProceduralShaders.h [hunk 5]
```diff
@@ -262,15 +262,39 @@ float snoise(vec2 v) {
...
+// shader playback time (in seconds)
+uniform float iGlobalTime;
+// the dimensions of the object being rendered
+uniform vec3 iWorldScale;            
 
 #define PROCEDURAL 1
 
 //PROCEDURAL_VERSION
+
+#ifdef PROCEDURAL_V1
```
Comment (ZappoMan) [inline libraries/procedural/src/procedural/ProceduralShaders.h]: what's this `#ifdef` all about?

**Pull request merged**